docs/phase_0.md

```diff
@@ -363,339 +363,1026 @@ class SignedVoluntaryExit(Container):
+        signing_root = compute_signing_root(signed_header.message, domain)
+        assert bls.Verify(proposer.pubkey, signing_root, signed_header.signature)
+
+    slash_validator(state, proposer_slashing.proposer_index)
+```
+
+It verifies the following validation conditions:
 
-<!insert_function `process_proposer_slashing`>
+* That the `slot` of `header_1` and `header_2` are equal
+* That `header_1` and `header_2` are not equal
+* That the proposer is slashable (`is_slashable_validator`). This means verifying that:
+    * The validator has not yet been `slashed` 
+    * The validator has been activated in the past -- i.e.`current_epoch` is greater than or equal to the validator `activation_epoch`
+    * The validator is not yet withdrawable -- i.e. `current_epoch` is less than the validator `withdrawable_epoch`
+* That both `header_1` and `header_2` have valid signatures from the validator defined by `proposer_index`
+
+```python
+def is_slashable_validator(validator: Validator, epoch: Epoch) -> bool:
+    """
+    Check if ``validator`` is slashable.
+    """
+    return (not validator.slashed) and (validator.activation_epoch <= epoch < validator.withdrawable_epoch)
+```
 
-<!insert_function `is_slashable_validator`>
+If the above conditions are met, `process_proposer_slashing` performs state updates to slash the `proposer`. This happens in the `slash_validator(state, proposer_index)` function call.
+
+
+```python
+def slash_validator(state: BeaconState,
+                    slashed_index: ValidatorIndex,
+                    whistleblower_index: ValidatorIndex=None) -> None:
+    """
+    Slash the validator with index ``slashed_index``.
+    """
+    epoch = get_current_epoch(state)
+    initiate_validator_exit(state, slashed_index)
+    validator = state.validators[slashed_index]
+    validator.slashed = True
+    validator.withdrawable_epoch = max(validator.withdrawable_epoch, Epoch(epoch + EPOCHS_PER_SLASHINGS_VECTOR))
+    state.slashings[epoch % EPOCHS_PER_SLASHINGS_VECTOR] += validator.effective_balance
+    decrease_balance(state, slashed_index, validator.effective_balance // MIN_SLASHING_PENALTY_QUOTIENT)
+
+    # Apply proposer and whistleblower rewards
+    proposer_index = get_beacon_proposer_index(state)
+    if whistleblower_index is None:
+        whistleblower_index = proposer_index
+    whistleblower_reward = Gwei(validator.effective_balance // WHISTLEBLOWER_REWARD_QUOTIENT)
+    proposer_reward = Gwei(whistleblower_reward // PROPOSER_REWARD_QUOTIENT)
+    increase_balance(state, proposer_index, proposer_reward)
+    increase_balance(state, whistleblower_index, whistleblower_reward - proposer_reward)
+```
 
 ##### Attester Slashings
 
-[`process_attester_slashing`](https://github.com/ethereum/eth2.0-specs/blob/v0.10.0/specs/phase0/beacon-chain.md#attester-slashings) defines the validation conditions that an `AttesterSlashing` must meet to b
```

Comment:
Remove extra space

```suggestion
- Even though the state transition may still be valid, the `Validator` is not allocated to the registry if the signature is invalid. Note that a user's signature may well be invalid, since it is not checked in the deposit contract.
```